rest_framework_json_api/schemas/openapi.py

```diff
@@ -0,0 +1,839 @@
+        """
+        Expand path containing .../{id}/{related_field} into list of related fields
+        and **their** views, making sure toOne relationship's views are a 'fetch' and toMany
+        relationship's are a 'list'.
+        :param path
+        :param method
+        :param view
+        :param view_endpoints
+        :return:list[tuple(path, method, view, action)]
+        """
+        result = []
+        serializer = view.get_serializer()
+        # It's not obvious if it's allowed to have both included_ and related_ serializers,
+        # so just merge both dicts.
+        serializers = {}
+        if hasattr(serializer, 'included_serializers'):
+            serializers = {**serializers, **serializer.included_serializers}
+        if hasattr(serializer, 'related_serializers'):
+            serializers = {**serializers, **serializer.related_serializers}
+        related_fields = [fs for fs in serializers.items()]
+
+        for field, related_serializer in related_fields:
+            related_view = self._find_related_view(view_endpoints, related_serializer, view)
+            if related_view:
+                action = self._field_is_one_or_many(field, view)
+                result.append(
+                    (path.replace('{related_field}', field), method, related_view, action)
+                )
+
+        return result
+
+    def _find_related_view(self, view_endpoints, related_serializer, parent_view):
+        """
+        For a given related_serializer, try to find it's "parent" view instance in view_endpoints.
+        :param view_endpoints: list of all view endpoints
+        :param related_serializer: the related serializer for a given related field
+        :param parent_view: the parent view (used to find toMany vs. toOne).
+        :return:view
+        """
+        for path, method, view in view_endpoints:
+            view_serializer = view.get_serializer()
+            if not isinstance(related_serializer, type):
+                related_serializer_class = import_class_from_dotted_path(related_serializer)
+            else:
+                related_serializer_class = related_serializer
+            if isinstance(view_serializer, related_serializer_class):
+                return view
+
+        return None
+
+    def _field_is_one_or_many(self, field, view):
+        serializer = view.get_serializer()
+        if isinstance(serializer.fields[field], ManyRelatedField):
+            return 'list'
+        else:
+            return 'fetch'
+
+
+class AutoSchema(drf_openapi.AutoSchema):
+    """
+    Extend DRF's openapi.AutoSchema for JSONAPI serialization.
+    """
+    content_types = ['application/vnd.api+json']
+
+    def __init__(self):
+        """
+        Initialize the JSONAPI OAS schema generator
+        """
+        super().__init__()
+
+    def get_operation(self, path, method, action=None):
+        """ basically a copy of AutoSchema.get_operation """
```

Comment:
basically a copy? When I look at the [upsteram implementation](https://github.com/encode/django-rest-framework/blame/4a98533746db44c997882ba01d5515a29a61dcc3/rest_framework/schemas/openapi.py#L141) it seems that it looks quite different. Could it be that this has heavily changed upstream since DJA work has been started on OAS?

We need to be careful copying code as it becomes a burden maintaining such code. It would be good to double check how this code can be best implemented and bring closer how DRF does it. And if overwriting the whole method is really necessary better to adjust this comment why overwriting is actually needed.

Reply:
yes, upstream changed a lot. He inlined something that had been a function call.

Reply:
comments update to reflect the fact this is not a copy, although, unfortunately, the current structure of upstream DRF still requires full replacements of several methods rather than simply extending them.